colossalai/pipeline/p2p.py

```diff
@@ -131,6 +146,224 @@ def _broadcast_object_list(
+            ops_queue.append(op_to_add)
+
+
+def create_recv_buffer(tensor_metadata, current_device):
+    if tensor_metadata[0] == 0:
+        tensor_shape, tensor_dtype, tensor_requires_grad = tensor_metadata[1]
+        tensor_recv = torch.empty(
+            tensor_shape, requires_grad=tensor_requires_grad, device=current_device, dtype=tensor_dtype)
+        return tensor_recv
+    elif tensor_metadata[0] == 1 or tensor_metadata[0] == 2:
+        buffer_recv = []
+        for tensor_data in tensor_metadata[1:]:
+            tensor_shape, tensor_dtype, tensor_requires_grad = tensor_data[-3:]
+            tensor_recv = torch.empty(
+                tensor_shape, requires_grad=tensor_requires_grad, device=current_device, dtype=tensor_dtype)
+            buffer_recv.append(tensor_recv)
+        return buffer_recv
+
+
+def _batch_send_recv_tensor(send_tensor_list, recv_tensor_metadata, send_dst, recv_src, send_group, recv_group, current_device):
+    buffer_recv = None
+    if recv_tensor_metadata is not None:
+        buffer_recv = create_recv_buffer(recv_tensor_metadata, current_device)
+
+    ops = []
+
+    if send_dst is not None:
+        filling_ops_queue(send_tensor_list, dist.isend, send_dst, ops, send_group)
+
+    if recv_src is not None:
+        assert buffer_recv is not None
+        filling_ops_queue(buffer_recv, dist.irecv, recv_src, ops, recv_group)
+
+    if len(ops) > 0:
+        reqs = dist.batch_isend_irecv(ops)
+        for req in reqs:
+            req.wait()
+    torch.cuda.synchronize()
+    return buffer_recv
+
+
+def _send_recv_serialization_object(
+        object: Any,
+        send_dst: Optional[int], recv_src: Optional[int],
+        send_group: Optional[ProcessGroup], recv_group: Optional[ProcessGroup],
+        current_device,
+        is_nccl_backend):
+    ops = []
+    send_object_tensor = None
+    if object is not None and send_dst is not None:
+        if Version(torch.__version__) >= Version("1.13.0"):
+            send_object_tensor, send_object_size_tensor = c10d._object_to_tensor(object, device=current_device)
+        else:
+            send_object_tensor, send_object_size_tensor = c10d._object_to_tensor(object)
+
+        if is_nccl_backend:
+            send_object_size_tensor = send_object_size_tensor.to(current_device)
+            send_object_tensor = send_object_tensor.to(current_device)
+
+        filling_ops_queue(send_object_size_tensor, dist.isend, send_dst, ops, send_group)
+
+    recv_object_size_tensor = None
+    if recv_src is not None:
+        recv_object_size_tensor = torch.empty(1, dtype=torch.long)
+        if is_nccl_backend:
+            recv_object_size_tensor = recv_object_size_tensor.to(current_device)
+        filling_ops_queue(recv_object_size_tensor, dist.irecv, recv_src, ops, recv_group)
+
+    if len(ops) > 0:
+        reqs = dist.batch_isend_irecv(ops)
+        for req in reqs:
+            req.wait()
+    torch.cuda.synchronize()
```

Comment:
I didn't find another stream. I think the synchronization is not necessary, right?